src/coreclr/tools/Common/Compiler/DependencyAnalysis/Relocation.cs

```diff
@@ -489,6 +489,45 @@ private static unsafe void PutRiscV64PC(uint* pCode, long imm32)
             Debug.Assert(GetRiscV64PC(pCode) == imm32);
         }
 
+        private static unsafe long GetRiscV64JALR(uint* pCode)
+        {
+            uint jalrInstr = *pCode;
+
+            // Extract and sign-extend the 12-bit signed immediate (bits 31:20 of the instruction).
+            return unchecked((int)jalrInstr) >> 20;
+        }
+
+        private static unsafe void PutRiscV64JALR(uint* pCode, long imm32)
+        {
+            // Verify that we got a valid offset within 32-bit signed range.
+            Debug.Assert((imm32 >= -0x80000000) && (imm32 < 0x80000000));
```

Comment:
That's true only if the 2nd instruction is 32-bit (addiw), jalr uses a 64-bit adder so the range is shifted.
```suggestion
            Debug.Assert((imm32 >= -0x80000000 - 0x800) && (imm32 < 0x80000000 - 0x800));
```
https://github.com/riscv/riscv-isa-manual/blob/9799e5b43fa6780e58ec440774b0194debbca52b/src/rv64.adoc?plain=1#L103-L105

(Please update in PutRiscV64PC).

Reply:
And the implementation of emitter::isValidSimm32 was also wrong as it's used only for auipc+jalr combos

Reply:
Thank you, consolidation is also working. Good idea to keep the types separate in JIT interface.